### PR TITLE
Fix: Improve the withdrawal packaging performance

### PR DESCRIPTION
**Pull request by jjyr**
* Fix a bug of withdrawal packaging. Before the fix, Godwoken skip packaging withdrawals every two blocks. Now, it packages withdrawals in every block.
* Increase maximum deposits & withdraws of each block from 50 to 100(100 depostis + 100 withdrawals costs ~ 240_000_000 cycles).
* Increase the maximum cycles of layer1 submission transaction to 350_000_000 (1/10 of a CKB block).
* Improve the logging of block-produce

### CODE REVIEW

File: crates/mem-pool/src/pool.rs [hunk 1]
```diff
@@ -914,11 +914,35 @@ impl MemPool {
...
+            // ignore withdrawal mismatch the nonce
+            nonce == expected_nonce
+        }
+
         // find withdrawals from pending
-        if withdrawals.is_empty() {
+        let mut withdrawals: Vec<_> = withdrawals
+            .into_iter()
+            .filter(|withdrawal| filter_withdrawals(state, withdrawal))
+            .collect();
```
Comment (blckngm) [inline crates/mem-pool/src/pool.rs]: Use [`retain`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.retain)?

File: crates/mem-pool/src/pool.rs [hunk 2]
```diff
@@ -914,11 +914,35 @@ impl MemPool {
...
+            withdrawal: &WithdrawalRequestExtra,
+        ) -> bool {
+            let id = state
+                .get_account_id_by_script_hash(&withdrawal.raw().account_script_hash().unpack())
+                .expect("get id")
+                .expect("id exist");
+            let nonce = state.get_nonce(id).expect("get nonce");
+            let expected_nonce: u32 = withdrawal.raw().nonce().unpack();
+            // ignore withdrawal mismatch the nonce
+            nonce == expected_nonce
```
Comment (keroro520) [inline crates/mem-pool/src/pool.rs]: What will happen IF `withdrawal.nonce < state.get_nonce(id)`? Is it possible?

Comment (keroro520) [reply crates/mem-pool/src/pool.rs]: Does the mem-pool check the nonce of a withdrawal when it receives one? If so, how does a withdrawal with a mismatched nonce enter the mem-pool?

**Comment by keroro520** [comment]
ACK the three items:
- Increase maximum deposits & withdraws of each block from 50 to 100.
- Increase the maximum cycles of layer1 submission transaction to 350_000_000 (1/10 of a CKB block).
- Improve the logging of block-produce

---

I do not quite understand how bug 659622294ddb7e89b8f664c011ee81396c40dd56 can be triggered. Can this bugfix be moved to another PR?

File: crates/mem-pool/src/pool.rs (hunk 2 above)
Comment (jjyr) [reply crates/mem-pool/src/pool.rs]: 1. It is not possible if our code work correctly.
2. When a new block arrived, mem-pool may still contains withdrawals in the new block, we remove these expired withdrawals in `MemPool#reset` function, that's the reason why we filter withdrawal here.

Comment (keroro520) [reply crates/mem-pool/src/pool.rs]: Now I get it. Thanks for explaining.

File: crates/config/src/config.rs [hunk 3]
```diff
@@ -221,7 +221,7 @@ fn default_enable_debug_rpc() -> bool {
 
 impl Default for DebugConfig {
     fn default() -> Self {
-        const EXPECTED_TX_UPPER_BOUND_CYCLES: u64 = 70000000u64;
+        const EXPECTED_TX_UPPER_BOUND_CYCLES: u64 = 350000000u64;
```
Comment (keroro520) [inline crates/config/src/config.rs:224]: Note that this limit update is the core fix.

**Review by keroro520 — DISMISSED** [review]
LGTM

File: crates/config/src/config.rs (hunk 3 above)
Comment (jjyr) [reply crates/config/src/config.rs:224]: The fix of withdrawal is also important, it double the count of packaged withdrawals in a block.

**Comment by jjyr** [comment]
> ACK the three items:
> 
> * Increase maximum deposits & withdraws of each block from 50 to 100.
> * Increase the maximum cycles of layer1 submission transaction to 350_000_000 (1/10 of a CKB block).
> * Improve the logging of block-produce
> 
> I do not quite understand how bug [6596222](https://github.com/nervosnetwork/godwoken/commit/659622294ddb7e89b8f664c011ee81396c40dd56) can be triggered. Can this bugfix be moved to another PR?

Updated the test case.

**Review by blckngm — DISMISSED** [review]

**Comment by zeroqn** [comment]
/itest
kicker: compatibility-changes

**Review by keroro520 — DISMISSED** [review]

**Comment by jjyr** [comment]
/itest
kicker: v1

**Comment by jjyr** [comment]
/itest
kicker: v1

**Comment by jjyr** [comment]
/itest
kicker: v1
web3: compatibility-breaking-changes

**Comment by jjyr** [comment]
/itest
kicker: compatibility-changes
web3: compatibility-breaking-changes

File: crates/tests/src/testing_tool/chain.rs [hunk 4]
```diff
@@ -446,15 +446,22 @@ pub async fn construct_block(
...
     chain: &Chain,
     mem_pool: &mut MemPool,
     deposit_requests: Vec<DepositRequest>,
     timestamp: u64,
+    refresh_mem_pool: bool,
 ) -> anyhow::Result<ProduceBlockResult> {
+    if !refresh_mem_pool {
+        assert!(
+            deposit_requests.is_empty(),
+            "skip refresh mem pool, bug deposits isn't empty"
```
Comment (blckngm) [inline crates/tests/src/testing_tool/chain.rs]: bug -> but?

**Review by zeroqn — DISMISSED** [review]

**Review by blckngm — APPROVED** [review]

**Pull request merged**